lib/postgrex/protocol.ex

```diff
@@ -702,7 +712,41 @@ defmodule Postgrex.Protocol do
 
     digest = :crypto.hash(:md5, [pass, user]) |> Base.encode16(case: :lower)
     digest = :crypto.hash(:md5, [digest, salt]) |> Base.encode16(case: :lower)
-    auth_send(s, msg_password(pass: ["md5", digest]), status, buffer)
+    auth_send(s, msg_password(pass: ["md5", digest, 0]), status, buffer)
+  end
+
+  defp auth_sasl(s, status = _, buffer) do
+    nonce = :crypto.strong_rand_bytes(div(@nonce_length*6, 8)) |> Base.encode64
+    message = IO.iodata_to_binary(["n,,n=,r=", nonce])
+    auth_send(s, msg_password(pass: ["SCRAM-SHA-256", 0, << byte_size(message) :: signed-size(32) >>, message]), status, buffer)
+  end
+
+  defp auth_cont(s, %{opts: opts} = status, data, buffer) do
+    kv_list = for kv <- String.split(data, ",") do
+        << k :: utf8, "=", v :: binary >> = kv
+        {k, v}
+      end
+    server = for {k, v} <- kv_list, into: %{}, do: {k, v}
+    client_nonce = String.slice(server[?r], 0, @nonce_length)
+
+    pass = Keyword.fetch!(opts, :password)
+    {server_i, ""} = Integer.parse(server[?i])
+    {:ok, server_s} = Base.decode64(server[?s])
+    salted_pass_hex = Pbkdf2.Base.hash_password(pass, server_s, [rounds: server_i, digest: :sha256, length: 32, format: :hex])
```

Comment:
I'm not sure how important is it, but PostgreSQL docs https://www.postgresql.org/docs/current/static/sasl-authentication.html#SASL-SCRAM-SHA-256 say that password should be "normalized" before hashing

> The SCRAM specification dictates that the password is also in UTF-8, and is processed with the SASLprep algorithm. PostgreSQL, however, does not require UTF-8 to be used for the password. When a user's password is set, it is processed with SASLprep as if it was in UTF-8, regardless of the actual encoding used. However, if it is not a legal UTF-8 byte sequence, or it contains UTF-8 byte sequences that are prohibited by the SASLprep algorithm, the raw password will be used without SASLprep processing, instead of throwing an error. This allows the password to be normalized when it is in UTF-8, but still allows a non-UTF-8 password to be used, and doesn't require the system to know which encoding the password is in.

In epgsql I decided to not implement all this complex normalization and only explicitly disallow non-ASCII in passwords: https://github.com/epgsql/epgsql/pull/156/files#diff-4a054884b454458eb3b45acded1fee2bR102